### PR TITLE
Add instrumentation for vert.x redis client

### PR DESCRIPTION
**Pull request by laurit**
This pr adds instrumentation for vert.x redis client. Instrumentation is added only for sending commands one by one, support for batch commands is not implemented.

### CODE REVIEW

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientTest.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.testcontainers.containers.GenericContainer;
+
+@SuppressWarnings("deprecation") // until old http semconv are dropped in 2.0
```
Comment (breedx-splk) [inline instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientTest.java:28]: We're beyond this now right?

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientInstrumentationModule.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    implements ExperimentalInstrumentationModule {
+
+  public VertxRedisClientInstrumentationModule() {
+    super("vertx-redis-client", "vertx-redis-client-4.0", "vertx");
+  }
+
+  @Override
+  public boolean isHelperClass(String className) {
+    return "io.vertx.redis.client.impl.RequestUtil".equals(className);
+  }
```
Comment (breedx-splk) [inline instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientInstrumentationModule.java:28]: Does the `RedisCommandSanitizer` also need to be declared as helper? It's used in the `VertxRedisClientAttributesGetter`...

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientSingletons.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+  public static void setCommandName(Command command, String commandName) {
+    commandNameField.set(command, commandName);
+  }
+
+  public static String getCommandName(Command command) {
+    return commandNameField.get(command);
+  }
+
+  private static final VirtualField<RedisStandaloneConnection, RedisURI> redisUriField =
```
Comment (breedx-splk) [inline instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientSingletons.java:98]: same comment about field being hidden.

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/RedisConnectionProviderInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  @SuppressWarnings("unused")
+  public static class InitAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnter(@Advice.FieldValue("redisURI") RedisURI redisUri) {
+      VertxRedisClientSingletons.setRedisUriThreadLocal(redisUri);
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void onExit() {
+      VertxRedisClientSingletons.setRedisUriThreadLocal(null);
```
Comment (breedx-splk) [inline instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/RedisConnectionProviderInstrumentation.java:48]: It's not immediately clear to me why 4.1.0 can use the ThreadLocal but the 4.0.0 init needs the VirtualField. Can we put a small comment about _why_?

(edit) I guess the comment in the StandAloneProviderAdvice kinda hints at this, but I was thinking about it up here.

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/vertx/redis/client/impl/RequestUtil.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.vertx.redis.client.impl;
```
Comment (breedx-splk) [inline instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/vertx/redis/client/impl/RequestUtil.java:6]: hacktastic! 🎈

**Review by breedx-splk — APPROVED** [review]
This looks really good to me. Nice work! I had a question about error spans, and a few small nits otherwise...seems mergeworthy to me.

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientInstrumentationModule.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+public class VertxRedisClientInstrumentationModule extends InstrumentationModule
+    implements ExperimentalInstrumentationModule {
+
+  public VertxRedisClientInstrumentationModule() {
+    super("vertx-redis-client", "vertx-redis-client-4.0", "vertx");
+  }
+
+  @Override
+  public boolean isHelperClass(String className) {
+    return "io.vertx.redis.client.impl.RequestUtil".equals(className);
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientInstrumentationModule.java:27]: Moving that class name to a constant won't really help. If you check other implementations of the same method then all of them are implemented pretty much like this.

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientTest.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                    span.hasName("RANDOMKEY")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(SemanticAttributes.DB_SYSTEM, "redis"),
+                            equalTo(SemanticAttributes.DB_STATEMENT, "RANDOMKEY"),
+                            equalTo(SemanticAttributes.DB_OPERATION, "RANDOMKEY"),
+                            equalTo(SemanticAttributes.NET_PEER_NAME, "localhost"),
+                            equalTo(SemanticAttributes.NET_PEER_PORT, port),
+                            equalTo(SemanticAttributes.NET_SOCK_PEER_PORT, port),
+                            equalTo(SemanticAttributes.NET_SOCK_PEER_ADDR, "127.0.0.1"))));
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientTest.java]: there is no error test, these tests were copied from jedis tests which don't have it either

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientSingletons.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+  public static RedisURI getRedisUriThreadLocal() {
+    return redisUriThreadLocal.get();
+  }
+
+  public static void setRedisUriThreadLocal(RedisURI redisUri) {
+    redisUriThreadLocal.set(redisUri);
+  }
+
+  private static final VirtualField<Command, String> commandNameField =
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientSingletons.java]: moved to top

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientRequest.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    return args;
+  }
+
+  public String getUser() {
+    return redisUri.user();
+  }
+
+  public String getDatabase() {
+    Integer select = redisUri.select();
+    return select != null ? select.toString() : null;
```
Comment (trask) [inline instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientRequest.java:41]: it looks like we could/should be using `db.redis.database_index` (https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/redis.md#semantic-conventions-for-redis)

**Review by trask — APPROVED** [review]

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientInstrumentationModule.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (breedx-splk) [reply instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/VertxRedisClientInstrumentationModule.java:27]: It's definitely a minor thing.

File: instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/RedisConnectionProviderInstrumentation.java (hunk 4 above)
Comment (breedx-splk) [reply instrumentation/vertx/vertx-redis-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/redis/RedisConnectionProviderInstrumentation.java:48]: Can we add a comment here as well tho?

**Pull request merged**